File: uber-rpm/du-customize/customize-pf9-installer/nettool.py
#!/bin/python
# Copyright (c) 2016 Platform9 Systems Inc.

import socket
import optparse

def resolve_fqdn(du_fqdn):
    try:
        socket.gethostbyname(du_fqdn)
        print("DNS resolution OK!")
    except socket.gaierror as e:
        print("Cannot resolve {0}: {1}".format(du_fqdn, e))
        exit(1)

def connect_to_host(du_fqdn, port):
    print("Connecting to {0}:{1}".format(du_fqdn, port))
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect((du_fqdn, port))
        sock.close()
        print("Connection to {0} succeeded!".format(du_fqdn))
    except socket.error as e:
        print("Error connecting to {0}: {1}".format(du_fqdn, e))
        exit(2)

def connect_to_host_via_proxy(proxy_host, proxy_port, du_fqdn, port):
    print("Connecting to {0}:{1}".format(proxy_host, proxy_port))
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect((proxy_host, proxy_port))
        print("Connection to {0}:{1} succeeded!".format(proxy_host,
                                                        proxy_port))
        print("Connecting to {0}:{1}".format(du_fqdn, port))
        sock.send("CONNECT {0}:{1} HTTP/1.0\r\n\r\n".format(du_fqdn,
                                                            port))
        res = sock.recv(1024)
        if "200 connection established" in res.lower():
            print("Connection to {0} via {1} succeeded!".format(du_fqdn,
                                                                proxy_host))
            sock.close()
        else:
            print("Error connecting to {0} via {1}:{2}".format(du_fqdn,
                                                             proxy_host,
                                                             proxy_port))
            sock.close()
            exit(3)
    except socket.error as e:
        print("Error connecting to {0}:{1}: {1}".format(proxy_host,
                                                        proxy_port,
                                                        e))
        exit(3)

def parse_options():
    usage="usage: %prog <command> [args]"
    parser = optparse.OptionParser(usage=usage)
    parser.add_option('--du-fqdn', action='store', dest='du_fqdn',
            help="Fully qualified domain name")

    parser.add_option('--port', action='store', dest='port',
            help="Port number")
    parser.add_option('--proxy-host', action='store', dest='proxy_host',
            help="Proxy host")

    parser.add_option('--proxy-port', action='store', dest='proxy_port',
            help="Port number")

    opts, args = parser.parse_args()
    return vars(opts), args

if __name__ == "__main__":
    options, args = parse_options()
    resolve_fqdn(options['du_fqdn'])
    if options['proxy_host'] is None:
        connect_to_host(options['du_fqdn'], int(options['port']))
    else:
        connect_to_host_via_proxy(options['proxy_host'],
                                  int(options['proxy_port']),
                                  options['du_fqdn'],
                                  int(options['port']))

File: uber-rpm/du-customize/customize-pf9-installer/nettool.go
/*
 *  Copyright (c) 2017 Platform9 Systems. All rights reserved
 *
 *  This is a Golang implementation of
 *  pf9-du/uber-rpm/du-customize/customize-pf9-installer/nettool.py
 *
 *  Please note that the binary is on the order of 3MB.
 *
 *  To link statically (no significant change in binary size):
 *  CGO_ENABLED=0 go build -installsuffix cgo nettool.go
 *
 */

package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"strings"
)

func dial (host string, port string) (net.Conn, error) {

	host_and_port := fmt.Sprintf ("%s:%s", host, port)
	log.Printf ("Connecting to %s", host_and_port)

	conn, err := net.Dial ("tcp", host_and_port)
	if err != nil {
		if strings.Contains (err.Error(), "no such host") {
			log.Printf ("Cannot resolve %s: %s", host, err)
			os.Exit (1)
		}
	}
	log.Print ("DNS resolution OK!")
	return conn, err
}

func connect_to_host (du_fqdn string, port string) int {

	conn, err := dial (du_fqdn, port)
	if err != nil {
		log.Printf ("Error connecting to %s: %s", du_fqdn, err)
	} else {
		log.Printf ("Connection to %s succeeded!", du_fqdn)
		conn.Close()
	}

	if err == nil {
		return 0
	}
	return 2
}

func connect_to_host_via_proxy (proxy_host string, proxy_port string, du_fqdn string, port string) int {

	success := false

	host_and_port := fmt.Sprintf ("%s:%s", proxy_host, proxy_port)

	conn, err := dial (proxy_host, proxy_port)
	if err != nil {
		log.Printf ("Error connecting to %s: %s", host_and_port, err)
	} else {
		log.Printf ("Connection to %s succeeded!", host_and_port)
		text := fmt.Sprintf ("CONNECT %s:%s HTTP/1.0\r\n\r\n", du_fqdn, port)
		fmt.Fprintf (conn, text)
		message, _ := bufio.NewReader(conn).ReadString('\n')
		if strings.Contains (strings.ToLower (message), "200 connection established") {
			log.Printf ("Connection to %s via %s:%s succeeded!", du_fqdn, proxy_host, proxy_port)
			success = true
		} else {
			log.Printf ("Error connecting to %s via %s:%s (%s)", du_fqdn, proxy_host, proxy_port, strings.TrimSpace (message))
		}
		conn.Close()
	}

	if success {
		return 0
	}
	return 3
}

func main() {

	du_fqdn := flag.String ("du-fqdn", "", "Fully qualified domain name")
	port := flag.String ("port", "443", "Port number")
	proxy_host := flag.String ("proxy-host", "", "Proxy host")
	proxy_port := flag.String ("proxy-port", "3128", "Proxy port number")

	flag.Parse()

	var success int

	if *proxy_host == "" {
		success = connect_to_host (*du_fqdn, *port)
	} else {
		success = connect_to_host_via_proxy (*proxy_host, *proxy_port, *du_fqdn, *port)
	}

	os.Exit (success)
}
